simulator/plugins/universalModule/main.cpp

```diff
@@ -0,0 +1,223 @@
+}
+
+std::string to_string(const gazebo::physics::JointWrench &jw)
+{
+    return "Forces:\n\tbody1: " + to_string(jw.body1Force) + "\n\tbody2: " + to_string(jw.body2Force)
+    + "\nTorques:\n\tbody1: " + to_string(jw.body1Torque) + "\n\tbody2: " + to_string(jw.body2Torque);
+}
+
+rofi::messages::RofiResp getJointRofiResp( rofi::messages::JointCmd::Type cmdtype, int joint, float value )
+{
+    rofi::messages::RofiResp resp;
+    resp.set_resptype( rofi::messages::RofiCmd::JOINT_CMD );
+    auto & jointResp = *resp.mutable_jointresp();
+    jointResp.set_joint( joint );
+    jointResp.set_cmdtype( cmdtype );
+
+    jointResp.add_values( value );
+
+    return resp;
+}
+
+using RofiCmdPtr = boost::shared_ptr< const rofi::messages::RofiCmd >;
+
+class UniversalModulePlugin : public ModelPlugin {
+public:
+
+    const double maxJointTorque = 10; // Max torque of joint motors
+    const std::pair< double, double > jointSpeedBoundaries = { 0, 10 }; // Max speed of joint motors
+    const std::pair< double, double > jointPositionBoundaries = { 0, 10 };
+
+    UniversalModulePlugin():
+        _pid( 0.1, 0, 0 ),
+        _node( boost::make_shared< transport::Node >() )
+    {}
+
+    virtual void Load( physics::ModelPtr model, sdf::ElementPtr sdf ) {
+        std::cerr << "\nThe UM plugin is attached to model ["
+                << model->GetName() << "]\n";
+
+        std::cerr << "Number of joints is " << model->GetJoints().size() << "\n";
+        _model = model;
+        auto _bodyJoint = _model->GetJoint( "bodyRev" );
+        auto _shoeAJoint = _model->GetJoint( "shoeARev" );
+        auto _shoeBJoint = _model->GetJoint( "shoeBRev" );
+        if ( !_bodyJoint || !_shoeAJoint || !_shoeBJoint ) {
+            std::cerr << "\nCould not get all joints\n";
+            return;
+        }
+
+        for ( auto &joint : { _shoeAJoint, _shoeBJoint } )
+        {
+            joint->SetParam( "vel", 0, 0.0 );
+            joint->SetParam( "fmax", 0, maxJointTorque );
+        }
+
+        _bodyJoint->SetParam( "vel", 0, 0.0 );
+        _bodyJoint->SetParam( "fmax", 0, maxJointTorque );
+
+        _node->Init( _model->GetWorld()->Name() );
+        std::string topicName = "~/" + _model->GetName() + "/control";
+        _sub = _node->Subscribe( topicName, &UniversalModulePlugin::onRofiCmd, this );
+        _pub = _node->Advertise< rofi::messages::RofiResp >( "~/" + _model->GetName() + "/response" );
+
+        std::cerr << "Listening...\n";
+    }
+private:
+    void onRofiCmd( const RofiCmdPtr &msg)
+    {
+        using rofi::messages::RofiCmd;
+
+        switch ( msg->cmdtype() )
+        {
+            case RofiCmd::NO_CMD:
+                break;
+            case RofiCmd::JOINT_CMD:
+                onJointCmd( msg->jointcmd() );
+                break;
+            case RofiCmd::DOCK_CMD:
+                std::cerr << "Dock commands not implemented\n";
```

Comment:
I would consider unifying the style of messages on cerr. Or maybe better - does Gazebo have a logger functionality? If so, I would use it.

Reply:
Used gazebo messages - in future it will be changed to gazebo logs